conference_scheduler/parameters.py

```diff
@@ -21,11 +50,46 @@ def _max_one_event_per_slot(shape, X):
         yield sum(X[(event, slot)] for event in range(shape.events)) <= 1
 
 
-def constraints(shape, X):
+def _slots_in_session(slot, session_array):
+    """
+    Return the indices of the slots in the same session as slot
+    """
+    return np.nonzero(session_array[slot])[0]
+
+def _events_with_diff_tag(talk, tag_array):
+    """
+    Return the indices of the events with no tag in common as tag
+    """
+    event_categories = np.nonzero(tag_array[talk])[0]
+    return np.nonzero(sum(tag_array.transpose()[event_categories]) == 0)[0]
+
+def _events_in_session_share_a_tag(session_array, tag_array, X):
+    """
+    Constraint that ensures that if a talk is in a given session then it must
+    share at least one tag with all other talks in that session.
+    """
+    event_indices = range(len(tag_array))
+    session_indices = range(len(session_array))
+    for session in session_indices:
+        slots = _slots_in_session(session, session_array)
+        for slot, event in itertools.product(slots, event_indices):
+            other_events = _events_with_diff_tag(event, tag_array)
+            for other_slot, other_event in itertools.product(slots, other_events):
+                if other_slot != slot and other_event != event:
+                    # If they have different tags they cannot be scheduled
+                    # together
+                    yield X[(event, slot)] + X[(other_event, other_slot)] <= 1
+
+
+def constraints(shape, session_array, tag_array, X):
     generators = (
         _schedule_all_events,
         _max_one_event_per_slot,
+        _events_in_session_share_a_tag,
     )
-    for generator in generators:
-        for constraint in generator(shape, X):
+    generator_kwargs = ({"shape":shape}, {"shape":shape},
```

Comment:
We have 3 constraint generators:

```python
 85     generators = (                                                                                                                                
 86         _schedule_all_events,                                                                                                                     
 87         _max_one_event_per_slot,                                                                                                                  
 88         _events_in_session_share_a_tag,                                                                                                           
 89     )    
```

Each of those is going to be called with (potentially different) kwargs:

```
 90     generator_kwargs = ({"shape":shape}, {"shape":shape},                                                                                         
 91                         {"session_array": session_array, "tag_array": tag_array}) 
```

That is done by zipping the two tuples together:

```python
93     for generator, kwargs in zip(generators, generator_kwargs):
```

So we need `shape` twice: for the first two generators.